### PR TITLE
fix(planner): reject subquery/agg/table in agg call

### LINKED ISSUE #4434: SqlSmith Bug hunt: Aggregate of Subquery results in panic
There is 1 out of 512 query that show this error which previous attempt to run 512 test that doesn,t fail. 

[error.txt](https://github.com/singularity-data/risingwave/files/9257514/error.txt)

it is error occur around here.
https://github.com/singularity-data/risingwave/blob/ae0388aeee470bb82b845e14df96220cf51e3d2a/src/frontend/src/expr/subquery.rs#L72
About hashing.

@xiangjinwu, Could you help me with this?

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

WIP: still polishing the code and testing for more cases.

Subquery / agg call / table function are all special expressions that may or may not be suitable to be part of group by / agg call inputs / agg call order by / agg call filter. For simplicity we reject all such queries as NotImplemented to avoid unexpected panics, even though some may be valid in PostgreSQL:

Note that the following is a valid query in PostgreSQL, although it looks like agg calls are nested (`min(min(a2) + b2)`). This is why we are changing from `InvalidInputSyntax` to `NotImplement` even for nested aggs.
```
select a1
from a
group by a1
having exists (
  select min(min(a2) + b2)
  from b
);
```

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

Fixes #4434

### CODE REVIEW

File: src/frontend/test_runner/tests/testdata/agg.yaml [hunk 1]
```diff
@@ -657,3 +659,106 @@
         LogicalProject { exprs: [t.x, t.y, flag] }
           LogicalExpand { column_subsets: [[t.x], [t.y]] }
             LogicalScan { table: t, columns: [t.x, t.y] }
+- sql: |
+    /* agg filter: subquery */
+    /* This case is valid in PostgreSQL */
+    create table a (a1 int, a2 int);
+    select count(a1) filter (where (select true)) from a;
+  binder_error: 'Invalid input syntax: subquery in filter clause is not supported'
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/agg.yaml:667]: Should we return NotImplemented in this case?

File: src/frontend/test_runner/tests/testdata/agg.yaml [hunk 2]
```diff
@@ -657,3 +659,106 @@
...
+- sql: |
+    /* agg filter: agg */
+    /* This case is valid in PostgreSQL */
+    create table a (a1 int, a2 int);
+    create table b (b1 int, b2 int);
+    select 1 from a having exists(
+      select count(b1) filter (where min(a1) < 3) from b
+    );
+  binder_error: 'Invalid input syntax: aggregation function in filter clause is not
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/agg.yaml:676]: ditto

**Review by xxchan — APPROVED** [review]
LGTM

File: src/frontend/test_runner/tests/testdata/agg.yaml (hunk 1 above)
Comment (xiangjinwu) [reply src/frontend/test_runner/tests/testdata/agg.yaml:667]: good catch

File: src/frontend/test_runner/tests/testdata/agg.yaml [hunk 3]
```diff
@@ -673,14 +678,17 @@
...
+    No tracking issue yet. Feel free to submit a feature request at https://github.com/risingwavelabs/risingwave/issues/new?labels=type%2Ffeature&template=feature_request.yml
 - sql: |
     /* agg filter: table function */
     /* This case is NOT valid in PostgreSQL */
     create table a (a1 int, a2 int);
     select count(a1) filter (where unnest(array[1]) < 1) from a;
-  binder_error: 'Invalid input syntax: table function in filter clause is not supported'
+  binder_error: |-
+    Feature is not yet implemented: table function in filter clause
+    No tracking issue yet. Feel free to submit a feature request at https://github.com/risingwavelabs/risingwave/issues/new?labels=type%2Ffeature&template=feature_request.yml
```
Comment (xxchan) [inline src/frontend/test_runner/tests/testdata/agg.yaml:691]: Table function is invalid instead of not implemented? 😂

Comment (xiangjinwu) [reply src/frontend/test_runner/tests/testdata/agg.yaml:691]: Based on my current test cases, yes. But I am not sure whether there are ways to construct a valid query that looks like "table function in filter clause" in the AST. Just like how we thought nested aggs are invalid, except that when the inner agg is correlated and belongs to the outer query.

**Pull request merged**